compiler/rustc_codegen_llvm/src/consts.rs

```diff
@@ -544,6 +549,165 @@ impl<'ll> CodegenCx<'ll, '_> {
     pub(crate) fn add_compiler_used_global(&mut self, global: &'ll Value) {
         self.compiler_used_statics.push(global);
     }
+
+    fn define_objc_classname(&self, classname: &str) -> &'ll Value {
+        // 32-bit x86 macOS only.
+        assert_eq!(self.tcx.sess.target.arch, "x86");
+        assert_eq!(self.tcx.sess.target.os, "macos");
+
+        let llval = self.null_terminate_const_bytes(classname.as_bytes());
+        let llty = self.val_ty(llval);
+        let sym = self.generate_local_symbol_name("OBJC_CLASS_NAME_");
+        let g = self.define_global(&sym, llty).unwrap_or_else(|| {
+            bug!("symbol `{}` is already defined", sym);
+        });
+        set_global_alignment(self, g, self.tcx.data_layout.i8_align.abi);
+        llvm::set_initializer(g, llval);
+        llvm::set_linkage(g, llvm::Linkage::PrivateLinkage);
+        llvm::set_section(g, c"__TEXT,__cstring,cstring_literals");
+        llvm::LLVMSetGlobalConstant(g, llvm::True);
+        llvm::LLVMSetUnnamedAddress(g, llvm::UnnamedAddr::Global);
+        g
+    }
+
+    fn get_objc_class_t(&self) -> &'ll Type {
+        if let Some(class_t) = self.objc_class_t.get() {
+            return class_t;
+        }
+
+        // Darwin-like targets other than 32-bit x86 macOS.
+        assert!(self.tcx.sess.target.is_like_darwin);
+        assert!(self.tcx.sess.target.arch != "x86" || self.tcx.sess.target.os != "macos");
+
+        let class_t = self.type_named_struct("struct._class_t");
+        let els = [self.type_ptr(); 5];
+        let packed = false;
+        self.set_struct_body(class_t, &els, packed);
+
+        self.objc_class_t.set(Some(class_t));
+        class_t
+    }
+
+    fn get_objc_classref(&self, classname: Symbol) -> &'ll Value {
+        let mut classrefs = self.objc_classrefs.borrow_mut();
+        if let Some(classref) = classrefs.get(&classname).copied() {
+            return classref;
+        }
+
+        let g = if self.tcx.sess.target.arch == "x86" && self.tcx.sess.target.os == "macos" {
+            // 32-bit x86 macOS.
+            let llval = self.define_objc_classname(classname.as_str());
+            let llty = self.type_ptr();
+            let sym = self.generate_local_symbol_name("OBJC_CLASS_REFERENCES_");
+            let g = self.define_global(&sym, llty).unwrap_or_else(|| {
+                bug!("symbol `{}` is already defined", sym);
+            });
+            set_global_alignment(self, g, self.tcx.data_layout.pointer_align().abi);
+            llvm::set_initializer(g, llval);
+            llvm::set_linkage(g, llvm::Linkage::PrivateLinkage);
```

Comment:
Q: In some places, you use `PrivateLinkage`, in others `InternalLinkage`. I assume this is intentional?